### ISSUE TITLE
Images with size markdown syntax break with remark

### ISSUE THREAD

**Issue by rogerhutchings** [zooniverse/front-end-monorepo#1269]
We need to support image rendering in answer buttons; Markdown isn't currently supported there due to a security feature in Remark, which is used in the new Markdownz component.

We can fix this by passing in the `img` element instead of using the `Media` component for image elements when rendering the Markdownz component inside the button.

**Comment by srallen** [comment]
Time estimate: 1 week, 1 person

**Comment by AnansiOmega** [comment]
Hey, I'm willing to take this on. But, a forewarning I'm new to contributing. I'd love some guidance if possible.

**Comment by AnansiOmega** [comment]
> Hey, I'm willing to take this on. But, a forewarning I'm new to contributing. I'd love some guidance if possible.

Just following up to see if you y'all are open to letting me take this on.

**Comment by srallen** [comment]
@AnansiOmega, sure! I've setup a project on staging demonstrating what the issue is: https://frontend.preview.zooniverse.org/projects/srallen086/galaxy-zoo-in-astronomy-101/classify?env=staging

The issue is that we're switching to using Remark as our markdown processor and the default configuration for it doesn't allow for anything but a valid URL in the URL portion of image markdown, e.g. `![image](https://www.example.com/image.png)`. This is for security reasons to prevent exploitations on XSS.

Our previous markdown processor, markdown-it, however uses a plugin called markdown-it-imsize which enabled support to specific widths and heights of images in markdown by adding the size into the URL area of the syntax, e.g. `![image](https://www.example.com/image.png =100x100)`.

We have a few options of how to move forward here:
1) Disable Remark's built in security and write our own that allows for this syntax. We've already rejected this option.
2) Require all projects using this syntax to remove their width and height specifications. This is not ideal, requires manual intervention, and would be time consuming.
3) Programmatically provide a backwards compatibility.

Option 3 is what we want to do. Our markdown processor is in our react component library. I had decided to add support for specifying the width and height in the alt text portion of the markdown syntax, e.g. `![image =100x100](https://example.com/image.png)`, which works (see: https://github.com/zooniverse/front-end-monorepo/blob/00d69241245cdade180f52fea7be70bc78ee7376/packages/lib-react-components/src/Markdownz/Markdownz.js#L52-L67). One approach could be to modify the function here to handle cleaning up the string for the URL, removing the size syntax and inputting it as props for the `Media` component as expected as well. This would require some additional regex parsing I think.

Another option would be what was described in the opening comment of the issue and to directly render an HTML image in the `TaskInput` component rather than parse the markdown, but I think this would still require regex parsing to figure out if we needed to this. I think I'd prefer updating the shared component so that we have a solution for this across our projects where markdown is used, not just in the task area.

We have specs and a storybook setup with the react component library to help with development. Let me know if you have any more questions! Thanks!

**Comment by AnansiOmega** [comment]
hey, @srallen I spent all of yesterday getting a lay of the land. As well as doing some research on Remark. One of the things I'm having trouble with is when I call nodeProps.src it returns a perfectly valid URL. I think the problem is that I'm opening up my development server using npm start. This opens up the storybook and throwing in debuggers in the function works fine too.

Furthermore, from my understanding the what you are looking for me to do is change this link:
 ``` ![Spiral_Image.png](https://panoptes-uploads.zooniverse.org/production/project_attached_image/0777d224-2e6f-410c-a5fe-885d4c873535.png =60x) Spiral ``` 
to look like this:
![Spiral_Image.png =60x](https://panoptes-uploads.zooniverse.org/production/project_attached_image/0777d224-2e6f-410c-a5fe-885d4c873535.png) Spiral

My apologies for the ignorance and please let me know if I'm missing something. Thank you!

**Comment by AnansiOmega** [comment]
![Spiral_Image.png =60x](https://panoptes-uploads.zooniverse.org/production/project_attached_image/0777-2e6f-410c-a5fe-885d4c873535.png) Spiral

( made an adjustment so you can see the url)

**Comment by AnansiOmega** [comment]
* ![Spiral_Image.png =60x](https://panoptes-uploads.zooniverse.org/production/project_attached_image/0777d224-2e6f-410c-a5fe-885d4c873535) Spiral

**Comment by srallen** [comment]
@AnansiOmega Yes, we want the image to render. The markdown `![Spiral_Image.png](https://panoptes-uploads.zooniverse.org/production/project_attached_image/0777d224-2e6f-410c-a5fe-885d4c873535.png =60x)` gets rendered as a string by Remark because it doesn't pass the built in security validation for the URL with the added size parameter using the previous markdown-it method. The stories and specs for `Markdownz` don't specifically test for this scenario, so first step would be to add a story and a spec that demonstrates the failure, then work to fix it to get them working and passing. The storybook stories use an example markdown file found in `/lib-react-components/.storybook/lib/example.md`. The specs use an example found in the component folder in `/helpers/testExamples.js`. Does this help? Do let me know if you have more questions!

**Comment by AnansiOmega** [comment]
Hey, @srallen I'd like to start off by asking, Is it ok if I ask to many questions? I've never worked with code at this level.
To begin, I've added in:
```
(line 20) ![a placeholder](https://panoptes-uploads.zooniverse.org/production/project_attached_image/0777d224-2e6f-410c-a5fe-885d4c873535.png =60x)

(line 106) <Paragraph>
    <img src="https://panoptes-uploads.zooniverse.org/production/project_attached_image/0777d224-2e6f-410c-a5fe-885d4c873535.png =60x" alt="a placeholder" />
  </Paragraph>
```

into the ```/helpers/testExamples.js``` folder, because of this addition I had to adjust Mardowns.spec.js file to accommodate this change:
```

  it('parses markdown to jsx', function () {
    wrapper = shallow(<Markdownz>{markdown}</Markdownz>)
    expect(wrapper.find(Anchor)).to.have.lengthOf(11)// prev 10
    expect(wrapper.find(Paragraph)).to.have.lengthOf(18)// prev 17
  })
```
Running yarn test gives me all tests are passing (yikes)

Furthermore, I also added in the old urls into the ```components/.storybook/lib/example.md```:
```
 |old image|`![Spiral_Image.png](https://panoptes-uploads.zooniverse.org/production/project_attached_image/0777d224-2e6f-410c-a5fe-885d4c873535.png)`|[![Spiral_Image.png](https://panoptes-uploads.zooniverse.org/production/project_attached_image/0777d224-2e6f-410c-a5fe-885d4c873535.png)|
|old image resized|`![Spiral_Image.png](https://panoptes-uploads.zooniverse.org/production/project_attached_image/0777d224-2e6f-410c-a5fe-885d4c873535.png =60x)`|![Spiral_Image.png](https://panoptes-uploads.zooniverse.org/production/project_attached_image/0777d224-2e6f-410c-a5fe-885d4c873535.png =60x)|
```

Now here come the questions..

- Do you want me to write a test for this url? ( If so, I'd like to let you know I've never written tests before so it'll rough but I'm willing to try )

- On the Markdownz.js file there is an Object called 'componentMappings' I'd like to think I understand what it's doing, but what I don't understand is where it's getting the 'nodeProps' from. My assumption is that it's taking this.props.children and parsing it out somehow?

- Within the renderMedia function I'm able to get the 'new urls' and they pass through, I'm also able to get the |old image| url. But when it comes to the | old image resized | the function completely ignores it and gives me the videos files. From what I understand is that there is something happening in between the componentMappings and nodeProps getting passed down to renderMedia. Also, I don't think Remark is able to get it's hands on this shotty URL yet because only later down the line does componentMappings gets passed through remark (line 105). So that can't be the thing 'filtering' out the bad URL. 


P.S - Thanks in advance, I find this an incredibly valuable learning experience.

**Comment by srallen** [comment]
@AnansiOmega 

> Running yarn test gives me all tests are passing (yikes)

This is expected until new tests are added to test for the missing new feature :) Remark's fallback isn't necessarily a failing condition unless we explicitly check for it. 

> Do you want me to write a test for this url? ( If so, I'd like to let you know I've never written tests before so it'll rough but I'm willing to try )

If you're up for it, then yes. This is a kind of approach to development to write a failing test first, then update the code to get it to pass and it gives us some assurance that the new feature is working as we think it should. Feel free to open a draft PR if you have questions on something specific about test writing. Having a bit of code to look at always helps when answering specific technical questions. The testing frameworks we're using is mocha, chai (for BDD style assertions), enzyme (React testing utility), and sinon.

> On the Markdownz.js file there is an Object called 'componentMappings' I'd like to think I understand what it's doing, but what I don't understand is where it's getting the 'nodeProps' from. My assumption is that it's taking this.props.children and parsing it out somehow?

The `componentMappings` object is for a function of the `remark-react` plugin (see: https://github.com/remarkjs/remark-react#optionsremarkreactcomponents). It normally defaults to render an HTML element by way of React JSX, but in our case, we want to render a specific React component for particular HTML elements. We can either specify a React component or supply a function that returns a React component. 

> Within the renderMedia function I'm able to get the 'new urls' and they pass through, I'm also able to get the |old image| url. But when it comes to the | old image resized | the function completely ignores it and gives me the videos files. From what I understand is that there is something happening in between the componentMappings and nodeProps getting passed down to renderMedia. Also, I don't think Remark is able to get it's hands on this shotty URL yet because only later down the line does componentMappings gets passed through remark (line 105). So that can't be the thing 'filtering' out the bad URL.

I tried this out locally, and you're right. This means that this will be more of a faff that I remembered it being. The security processing by remark is happening before the component mapping from remark-react happens. We could try to parse `this.props.children` which is the string of markdown before it even gets processed by remark, but that seems really hairy to do 😬 . I've noticed that we're a few major versions behind, though, and I read in the current docs that it seems the internals of the remark processor has changed, so maybe first step instead could be to update the remark and remark plugin versions (using `yarn upgrade-interactive --latest` at the root of the repository makes this easy to do) in a separate PR and then we go from there. Maybe there's a better way to handle this with the updated version. 

Also one last thing, with end of year holiday time off approaching soon, responses may slow down, so do know that we do appreciate your contribution, however, our responses to issues and PRs may be slower this time a year and will pick back up in January. Thanks!
